packages/safe-ds-vscode/src/extension/pythonServer.ts

```diff
@@ -0,0 +1,334 @@
+let pythonServerConnection: WebSocket | undefined = undefined;
+let pythonServerMessageCallbacks: Map<PythonServerMessage['type'], ((message: PythonServerMessage) => void)[]> =
+    new Map<PythonServerMessage['type'], ((message: PythonServerMessage) => void)[]>();
+
+/**
+ * Start the python server on the next usable port, starting at 5000.
+ * Uses the 'safe-ds.runner.command' setting to execute the process.
+ */
+export const startPythonServer = async function (): Promise<void> {
+    pythonServerAcceptsConnections = false;
+    const runnerCommandSetting = vscode.workspace.getConfiguration('safe-ds.runner').get<string>('command')!; // Default is set
+    pythonServerPort = await findFirstFreePort(5000);
+    logOutput(`Trying to use port ${pythonServerPort} to start python server...`);
+    logOutput(`Using command '${runnerCommandSetting}' to start python server...`);
+    const runnerCommandParts = runnerCommandSetting.split(/\s/u);
+    const runnerCommand = runnerCommandParts.shift()!;
+    const runnerArgs = [...runnerCommandParts, '--port', String(pythonServerPort)];
+    logOutput(`Running ${runnerCommand}; Args: ${runnerArgs.join(' ')}`);
+    pythonServer = child_process.spawn(runnerCommand, runnerArgs);
+    manageSubprocessOutputIO(pythonServer);
+    await connectToWebSocket();
+    logOutput('Started python server successfully');
+};
+
+export const stopPythonServer = function (): void {
+    pythonServer?.kill();
+    // TODO maybe tree-kill sync process? Does not seem to always be reaped, and may remain as a zombie
+    pythonServer = undefined;
+};
+
+/**
+ * @return True if the python server was started and the websocket connection was established, false otherwise.
+ */
+export const isPythonServerAvailable = function (): boolean {
+    return pythonServerAcceptsConnections;
+};
+
+/**
+ * Register a callback to execute when a message from the python server arrives.
+ *
+ * @param callback Callback to execute
+ * @param messageType Message type to register the callback for.
+ */
+export const addMessageCallback = function <M extends PythonServerMessage['type']>(
+    callback: (message: Extract<PythonServerMessage, { type: M }>) => void,
+    messageType: M,
+): void {
+    if (!pythonServerMessageCallbacks.has(messageType)) {
+        pythonServerMessageCallbacks.set(messageType, []);
+    }
+    pythonServerMessageCallbacks.get(messageType)!.push(<(message: PythonServerMessage) => void>callback);
+};
+
+const importPipeline = async function (services: SafeDsServices, documentUri: URI): Promise<LangiumDocument> {
+    const document = services.shared.workspace.LangiumDocuments.getOrCreateDocument(documentUri);
+    await services.shared.workspace.DocumentBuilder.build([document], { validation: true });
+
+    const errors = (document.diagnostics ?? []).filter((e) => e.severity === 1);
+
+    if (errors.length > 0) {
+        let errorString = `The document ${documentUri.fsPath} has errors:`;
```

Comment:
For logging, you should always favor uri.toString() over uri.fsPath. The former is rendered as a clickable link by VS Code, the latter is not.